sql: expose metrics snapshot in crdb_internal

tbg opened the pull request:
This includes a metrics snapshot in crdb_internal.node_metrics, and also
scrapes this table as part of `debug zip`.

In follow-up work, we can from these metrics extract a health check that
operates on the metric snapshot. For example, you might check whether
there are underreplicated ranges, or high heartbeat latencies. Taking
this one step further, we may introduce a periodic query on this table
as a health check, with a distress signal sent through gossip when
problems are detected.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25308)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25308#-:-LBqqPGJ4I61AoF1Tu_a:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
sql: expose metrics snapshot in crdb_internal

This includes a metrics snapshot in crdb_internal.node_metrics, and also
scrapes this table as part of `debug zip`.

In follow-up work, we can from these metrics extract a health check that
operates on the metric snapshot. For example, you might check whether
there are underreplicated ranges, or high heartbeat latencies. Taking
this one step further, we may introduce a periodic query on this table
as a health check, with a distress signal sent through gossip when
problems are detected.

Release note: None

tschottdorf commented:
TFTR!

bors r=bdarnell

a-robinson commented on pkg/sql/crdb_internal.go:
```diff
@@ -867,6 +868,46 @@ func populateSessionsTable(
+
+		mr := p.ExecCfg().MetricsRecorder
+		if mr == nil {
+			return nil
```
Shouldn't we return an error if we're unable to return the requested information?

a-robinson commented on pkg/sql/crdb_internal.go:
```diff
@@ -867,6 +868,46 @@ func populateSessionsTable(
+// current node.
+var crdbInternalLocalMetricsTable = virtualSchemaTable{
+	schema: `CREATE TABLE crdb_internal.node_metrics (
+  store_id 	         INT NULL,       -- the store, if any, to which this metric belongs
```
Should we also include a `node_id` column so that it's clear when connecting through a load balancer which node the metrics are from? Many other `crdb_internal` tables that only report local node information do so ( e.g. `node_build_info`, `leases`, `node_statement_statistics`, etc).

a-robinson reviewed (commented):
Sorry for not checking this out on Friday!

petermattis commented:
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/debug.go, line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25308#-LD7fcQQ59zvhNUynvSW:-LD7fcQQ59zvhNUynvSX:b-ysnci1) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e0784b5010e4ea9ab537c389572f4c8a60d77a/pkg/cmd/roachtest/debug.go#L58)):*
> ```Go
> 			if err := c.RunE(ctx, c.Node(node), "grep -F 'liveness.heartbeatlatency-p99' ./debug/metrics"); err != nil {
> 				return err
> 			}
> ```

@tschottdorf I've arrived here via #25804. Do we want to set `MinVersion` for this test to `v2.1.0`? That seems unfortunate. Should I add a mechanism so that a test can optionally perform additional tests if the build version is at some value? Something like `test.IsBuildVersion(string)`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25308)*
<!-- Sent from Reviewable.io -->
